pkg/pool-weighted/test/ManagedPoolTokenLib.test.ts

```diff
@@ -0,0 +1,169 @@
+import { expect } from 'chai';
+import { Contract, BigNumber, Wallet } from 'ethers';
+import { hexlify, randomBytes } from 'ethers/lib/utils';
+import { deploy } from '@balancer-labs/v2-helpers/src/contract';
+import { BigNumberish, bn, fp, negate } from '@balancer-labs/v2-helpers/src/numbers';
+import { random, range } from 'lodash';
+import Token from '@balancer-labs/v2-helpers/src/models/tokens/Token';
+import { toNormalizedWeights } from '@balancer-labs/balancer-js';
+
+describe('ManagedPoolTokenLib', () => {
+  let lib: Contract;
+
+  const MAX_RELATIVE_ERROR = 0.0005;
+  const TEST_RUNS = 10;
+
+  before('deploy lib', async () => {
+    lib = await deploy('MockManagedPoolTokenLib');
+  });
+
+  function checkMaskedWord(result: string, word: string, offset: number, bits: number): void {
+    // Mask to keep all bits outside of the length of bits `bits` at offset `offset`.
+    const mask = negate(bn(1).shl(bits).sub(1).shl(offset));
+
+    // All masked bits should match the original word.
+    const clearedResult = mask.and(result);
+    const clearedWord = mask.and(word);
+    expect(clearedResult).to.equal(clearedWord);
+  }
+
+  describe('token scaling factor', () => {
+    const DECIMAL_DIFF_OFFSET = 128;
+    const DECIMAL_DIFF_WIDTH = 5;
+
+    async function assertTokenScalingFactor(
```

Comment:
I'd have this function also receive an `expected` argument, and also check against that.

Reply:
As in an expected value for `result`? I'm inclined to work just off of observable values instead as this is all the Pool cares about (the pool should only use `_tokenState` to feed it into this lib).

Reply:
This would likely get pretty yucky as we'd have to be able to predict the rounding errors involved in compressing the weights as we can't do fuzzy matching on the entire state without decoding it first (at which point we'd just be back to the current testing regime but we're testing everything in every test.)